### PR TITLE
Add Erigon support for devnets

### PR DESCRIPTION
**Pull request by Raneet10**
# Description

This PR adds support for running erigon client in a **remote** PoS devnet, including as a validator (feature to run public node is yet to be enabled). Note the express-cli commands/tests are yet to be updated.

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] New test case for remote devnet

# Breaking changes

Some config variables are added and/or are removed from the `.env`.

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

## Testing

- [ ] I have tested this code manually on local environment
- [x] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai

### CODE REVIEW

File: src/express/common/config-utils.js [hunk 1]
```diff
@@ -480,10 +595,55 @@ export async function editMaticCliRemoteYAMLConfig() {
...
-  setConfigList('devnetBorHosts', process.env.DEVNET_BOR_HOSTS, doc)
-  setConfigList('devnetHeimdallHosts', process.env.DEVNET_BOR_HOSTS, doc)
-  setConfigList('devnetBorUsers', process.env.DEVNET_BOR_USERS, doc)
-  setConfigList('devnetHeimdallUsers', process.env.DEVNET_BOR_USERS, doc)
+  if (!process.env.DEVNET_BOR_USERS) {
+    setConfigList('devnetErigonHosts', process.env.DEVNET_ERIGON_HOSTS, doc)
+    setConfigList('devnetErigonUsers', process.env.DEVNET_ERIGON_USERS, doc)
+    setConfigList('devnetHeimdallUsers', process.env.DEVNET_ERIGON_USERS, doc)
+    setConfigList('devnetHeimdallHosts', process.env.DEVNET_ERIGON_HOSTS, doc)
+    deleteConfig('devnetBorUsers', doc)
```
Comment (marcello33) [inline src/express/common/config-utils.js:603]: Why do we set earlier if we know we'll delete afterwards?
This whole process was meant to have bor only, but maybe now it's about time to refactor it in a way it's suitable for more clients, without too many complications. Maybe we can discuss new approaches over a call

File: src/express/common/config-utils.js [hunk 2]
```diff
@@ -498,6 +658,53 @@ export async function editMaticCliRemoteYAMLConfig() {
   )
 }
 
+export function setBorAndErigonHosts(dnsIps) {
```
Comment (marcello33) [inline src/express/common/config-utils.js:661]: This - as before - is getting too complicated. Let's sit and find a better way

File: src/setup/devnet/index.js [hunk 3]
```diff
@@ -421,7 +516,27 @@ export class Devnet {
...
+            chaindata = '~/.bor/data/bor/chaindata'
+            if (i >= this.config.numOfBorValidators && erigonValCount > 0) {
+              host = this.config.devnetErigonHosts[i - this.config.numOfNonBorValidators]
+              user = this.config.devnetErigonUsers[i - this.config.numOfBorValidators]
+              snapshotUrl = this.config.erigonSnapshotUrl
+              service = 'erigon.service'
+              chaindata = '~/.erigon/data/erigon/chaindata'
+            }
+            if (i >= this.totalBorNodes + this.config.numOfErigonValidators) {
+              host = this.config.devnetErigonHosts[i - (this.totalBorNodes + this.config.numOfErigonValidators)]
```
Comment (marcello33) [inline src/setup/devnet/index.js:534]: Why is this part duplicated on a different condition?

**Review by marcello33 — CHANGES_REQUESTED** [review]

File: main.tf [hunk 4]
```diff
@@ -16,20 +16,40 @@ provider "aws" {
...
+    volume_type = (count.index >= var.BOR_VALIDATOR_COUNT + var.BOR_SENTRY_COUNT) ? var.BOR_ARCHIVE_VOLUME_TYPE : var.BOR_VOLUME_TYPE
+    iops = (count.index >= var.BOR_VALIDATOR_COUNT + var.BOR_SENTRY_COUNT ) ? var.BOR_ARCHIVE_IOPS : var.BOR_IOPS
+  }
+
+  tags = {
+    Name = "${var.VM_NAME}_${count.index + 1}"
+  }
+}
+resource "aws_instance" "erigon_node_server" {
+  count = (var.DOCKERIZED == "yes") ? 1 : (var.ERIGON_VALIDATOR_COUNT + var.ERIGON_SENTRY_COUNT + var.ERIGON_ARCHIVE_COUNT)
```
[the remark this answers is not in the document]
Comment (Raneet10) [reply main.tf:40]: Yeah I see your point. I actually targeted the remote setup but ended up adding the condition for docker as well; it's probably not needed in case of Erigon. Let me know your thoughts.

[the remark this answers is not in the document]
Comment (marcello33) [reply main.tf:40]: If we remove it, then we can never run a dockerized env with erigon only
But if we leave it, we'll create 2 servers which is an error. Maybe we need to move/rethink this check

File: .env.example [hunk 5]
```diff
@@ -1,26 +1,37 @@
...
-TF_VAR_VOLUME_TYPE=gp3 # default gp3
-TF_VAR_ARCHIVE_VOLUME_TYPE=io1 # type of EBS volume for archive nodes (default is io1)
-TF_VAR_INSTANCE_AMI=ami-017fecd1353bcc96e# ami instance type (default is ami-017fecd1353bcc96e, running ubuntu 22.04)
-TF_VAR_PEM_FILE=aws-key# name of the certificate (.pem file) previously generated (default is aws-key)
-TF_VAR_REGION=us-west-2# AWS region to deploy to (default is us-west-2)
-PEM_FILE_PATH=/absolute/path/to/your/aws-key.pem# absolute path pointing to the certificate previously downloaded
+TF_VAR_AWS_PROFILE=default # do not change this
+TF_VAR_VM_NAME=YOUR_IDENTIFIER # default "polygon-user". It can be any string, used to discriminate between instances
+TF_VAR_DOCKERIZED=no # default "no", otherwise only one VM is created and the Polygon devnet will run in docker containers
+TF_VAR_BOR_DISK_SIZE_GB=100 # size of the disk in GB for Bor client (default is 100GB)
```
Comment (anshalshukla) [inline .env.example:5]: We were planning on reducing this to 20, can we do this here or a separate PR is required?

File: .env.example [hunk 6]
```diff
@@ -1,26 +1,37 @@
...
+TF_VAR_ERIGON_ARCHIVE_DISK_SIZE_GB=100 # size of the disk in GB in Erigon archive node (default is 100GB)
+TF_VAR_BOR_IOPS=3000 # Amount of provisioned IOPS for Bor client
+TF_VAR_ERIGON_IOPS=3000 # Amount of provisioned IOPS for Erigon client
+TF_VAR_BOR_ARCHIVE_IOPS=15000 # Amount of provisioned IOPS in Bor archive node
+TF_VAR_ERIGON_ARCHIVE_IOPS=15000 # Amount of provisioned IOPS in Erigon archive node
+TF_VAR_BOR_VALIDATOR_COUNT=2 # number of Bor validator nodes (default is 2). Note that while spinning up a public network (mainnet/mumbai) node, this will serve as a non-validator
+TF_VAR_BOR_SENTRY_COUNT=1 # number of non-validator Bor sentry nodes (default is 1)
+TF_VAR_BOR_ARCHIVE_COUNT=0 # number of Bor archive nodes (default is 0)
+TF_VAR_ERIGON_VALIDATOR_COUNT=0 #number of Erigon validator nodes (default is 0). Note that while spinning up a public network (mainnet/mumbai) node, this will serve as a non-validator
+TF_VAR_ERIGON_SENTRY_COUNT=0 #number of non-validator Erigon sentry nodes (default is 0)
```
Comment (anshalshukla) [inline .env.example:17]: Same

File: .env.example [hunk 7]
```diff
@@ -30,37 +41,41 @@ SPRINT_SIZE_BLOCK_NUMBER=0 # comma separated values defining the block heights o
...
 CONTRACTS_REPO="https://github.com/maticnetwork/contracts.git" # repo of contracts project
 CONTRACTS_BRANCH=mardizzone/node-16 # contracts branch defining the version to be used in the devnet. To use node v16.x, please stick with it
 GENESIS_CONTRACTS_REPO="https://github.com/maticnetwork/genesis-contracts.git" # repo of genesis-contracts project
 GENESIS_CONTRACTS_BRANCH=master# Genesis contracts branch defining the version to be used in the devnet
 MATIC_CLI_REPO="https://github.com/maticnetwork/matic-cli.git" # repo of matic-cli to run a specific version remotely using express-cli
-MATIC_CLI_BRANCH=master# matic-cli branch used on the remote machines to start the environment
-DEVNET_BOR_USERS=ubuntu,ubuntu,ubuntu#users' names of VMs for all the nodes (comma separated). Its length must be equal to "TF_VAR_VALIDATOR_COUNT + TF_VAR_SENTRY_COUNT"
+MATIC_CLI_BRANCH=master # matic-cli branch used on the remote machines to start the environment
+DEVNET_BOR_USERS=ubuntu,ubuntu,ubuntu #users' names of VMs for all the nodes (comma separated). Its length must be equal to "TF_VAR_BOR_VALIDATOR_COUNT + TF_VAR_BOR_SENTRY_COUNT + TF_VAR_BOR_ARCHIVE_COUNT"
```
Comment (anshalshukla) [inline .env.example:55]: Spacing 😅

File: .env.example [hunk 8]
```diff
@@ -30,37 +41,41 @@ SPRINT_SIZE_BLOCK_NUMBER=0 # comma separated values defining the block heights o
...
+DEVNET_ERIGON_USERS=ubuntu #users' names of VMs for all the nodes (comma separated). Its length must be equal to "TF_VAR_ERIGON_VALIDATOR_COUNT + TF_VAR_ERIGON_SENTRY_COUNT + TF_VAR_ERIGON_ARCHIVE_COUNT"
```
Comment (anshalshukla) [inline .env.example:56]: spacing

File: .env.example [hunk 9]
```diff
@@ -30,37 +41,41 @@ SPRINT_SIZE_BLOCK_NUMBER=0 # comma separated values defining the block heights o
...
-VERBOSE=true# if set to true will print logs also from remote machines
-DD_API_KEY=DATADOG_API_KEY# Datadog API key
+VERBOSE=true # if set to true will print logs also from remote machines
+DD_API_KEY=DATADOG_API_KEY # Datadog API key
 
-# Bor and Heimdall Snapshot related variables (when running mainnet/testnet node)
+# Bor, Erigon and Heimdall Snapshot related variables (when running mainnet/testnet node)
 BOR_SNAPSHOT_URL=https://snapshot-download.polygon.technology/bor-mumbai-fullnode-2023-03-18.tar.zst # URL for bor data snapshot (refer: https://snapshot.polygon.technology/)
 HEIMDALL_SNAPSHOT_URL=https://snapshot-download.polygon.technology/heimdall-mumbai-fullnode-2023-03-18.tar.zst # URL for heimdall data snapshot (refer: https://snapshot.polygon.technology/)
+ERIGON_SNAPSHOT_URL= # URL for Erigon data snapshot (refer: https://snapshot.polygon.technology/)
```
Comment (anshalshukla) [inline .env.example:65]: Maybe you can put this as default  https://snapshot-download.polygon.technology/erigon-mumbai-archive-2023-05-24.tar.zst

File: src/setup/devnet/templates/remote/erigon-service-host.sh [hunk 10]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env sh
+
+NODE_DIR=$HOME/node
+BOR_HOME=$HOME/.bor
```
Comment (anshalshukla) [inline src/setup/devnet/templates/remote/erigon-service-host.sh:4]: ERIGON_HOME=$HOME/.erigon?

File: src/setup/devnet/templates/remote/erigon-service.sh [hunk 11]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env sh
+
+NODE_DIR=$HOME/node
+BOR_HOME=$HOME/.bor
```
Comment (anshalshukla) [inline src/setup/devnet/templates/remote/erigon-service.sh:4]: ERIGON_HOME?

File: variables.tf [hunk 12]
```diff
@@ -8,49 +8,93 @@ variable "VM_NAME" {
   default = "polygon-user"
 }
 
-variable "DISK_SIZE_GB" {
+variable "BOR_DISK_SIZE_GB" {
   default = "100"
```
Comment (anshalshukla) [inline variables.tf:12]: Same can we change default to 20?

File: variables.tf [hunk 13]
```diff
@@ -8,49 +8,93 @@ variable "VM_NAME" {
...
+variable "ERIGON_ARCHIVE_COUNT" {
   default = "0"
 }
-variable "INSTANCE_TYPE" {
+variable "BOR_INSTANCE_TYPE" {
+  default = "t2.xlarge"
+}
+
+variable "ERIGON_INSTANCE_TYPE" {
+  default = "t2.xlarge"
```
Comment (anshalshukla) [inline variables.tf:90]: I think we should have r5b.large in case of erigon, generally speaking erigon nodes perform better on memory optimised instances and r5b.large is cheaper than t2.xl

File: variables.tf [hunk 14]
```diff
@@ -8,49 +8,93 @@ variable "VM_NAME" {
...
+}
+
+variable "BOR_ARCHIVE_INSTANCE_TYPE" {
   default = "t2.xlarge"
 }
 
-variable "ARCHIVE_INSTANCE_TYPE" {
+variable "ERIGON_ARCHIVE_INSTANCE_TYPE" {
   default = "t2.xlarge"
```
Comment (anshalshukla) [inline variables.tf:98]: r5b.large

File: main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (anshalshukla) [reply main.tf:40]: Can we have a separate resource block to handle dockerized cases like we are currently handling bor and erigon in separate resource blocks so does a third resource for docker makes sense?

File: .env.example [hunk 15]
```diff
@@ -1,26 +1,37 @@
...
+TF_VAR_AWS_PROFILE=default # do not change this
+TF_VAR_VM_NAME=YOUR_IDENTIFIER # default "polygon-user". It can be any string, used to discriminate between instances
+TF_VAR_DOCKERIZED=no # default "no", otherwise only one VM is created and the Polygon devnet will run in docker containers
+TF_VAR_BOR_DISK_SIZE_GB=20 # size of the disk in GB for Bor client (default is 100GB)
+TF_VAR_ERIGON_DISK_SIZE_GB=20 # size of the disk in GB for Erigon client (default is 100GB)
+TF_VAR_BOR_ARCHIVE_DISK_SIZE_GB=100 # size of the disk in GB in Bor archive node (default is 100GB)
+TF_VAR_ERIGON_ARCHIVE_DISK_SIZE_GB=100 # size of the disk in GB in Erigon archive node (default is 100GB)
+TF_VAR_BOR_IOPS=3000 # Amount of provisioned IOPS for Bor client
+TF_VAR_ERIGON_IOPS=3000 # Amount of provisioned IOPS for Erigon client
+TF_VAR_BOR_ARCHIVE_IOPS=15000 # Amount of provisioned IOPS in Bor archive node
```
Comment (anshalshukla) [inline .env.example:11]: I was just trying with lower sized disk and we cannot have high IOPS with such small disks 😅

File: variables.tf [hunk 16]
```diff
@@ -8,52 +8,96 @@ variable "VM_NAME" {
...
+variable "BOR_DISK_SIZE_GB" {
+  default = "20"
+}
+
+variable "ERIGON_DISK_SIZE_GB" {
+  default = "20"
+}
+
+variable "BOR_ARCHIVE_DISK_SIZE_GB" {
   default = "100"
```
Comment (anshalshukla) [inline variables.tf:20]: 20?

File: variables.tf [hunk 17]
```diff
@@ -8,52 +8,96 @@ variable "VM_NAME" {
...
 }
 
-variable "ARCHIVE_DISK_SIZE_GB" {
+variable "ERIGON_ARCHIVE_DISK_SIZE_GB" {
   default = "100"
```
Comment (anshalshukla) [inline variables.tf:24]: 20?

File: .env.example (hunk 15 above)
Comment (anshalshukla) [reply .env.example:11]: I'll have to check for the limit, 1000 worked for me but I'm not sure about the max/prescribed limit

**Review by anshalshukla — CHANGES_REQUESTED** [review]

File: main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (Raneet10) [reply main.tf]: Done here: https://github.com/maticnetwork/matic-cli/pull/184/commits/858c3f213f94d97deaa9cfaf9cef0c0da661f7e7

**Review by anshalshukla — APPROVED** [review]
